pkg/proc/target_exec.go

```diff
@@ -211,39 +282,103 @@ func isTraceOrTraceReturn(bp *Breakpoint) bool {
 	return bp.Logical.Tracepoint || bp.Logical.TraceReturn
 }
 
-func conditionErrors(threads []Thread) error {
+func conditionErrors(grp *TargetGroup) error {
 	var condErr error
-	for _, th := range threads {
-		if bp := th.Breakpoint(); bp.Breakpoint != nil && bp.CondError != nil {
-			if condErr == nil {
-				condErr = bp.CondError
-			} else {
-				return fmt.Errorf("multiple errors evaluating conditions")
+	for _, dbp := range grp.targets {
+		if isvalid, _ := dbp.Valid(); !isvalid {
+			continue
+		}
+		for _, th := range dbp.ThreadList() {
+			if bp := th.Breakpoint(); bp.Breakpoint != nil && bp.CondError != nil {
+				if condErr == nil {
+					condErr = bp.CondError
+				} else {
+					return fmt.Errorf("multiple errors evaluating conditions")
+				}
 			}
 		}
 	}
 	return condErr
 }
 
 // pick a new dbp.currentThread, with the following priority:
+//
 //   - a thread with an active stepping breakpoint
 //   - a thread with an active breakpoint, prioritizing trapthread
-//   - trapthread
-func pickCurrentThread(dbp *Target, trapthread Thread, threads []Thread) error {
+//   - trapthread if it is not nil
+//   - the previous current thread if it still exists
+//   - a randomly selected thread
+func pickCurrentThread(dbp *Target, trapthread Thread) error {
+	threads := dbp.ThreadList()
 	for _, th := range threads {
 		if bp := th.Breakpoint(); bp.Active && bp.Stepping {
 			return dbp.SwitchThread(th.ThreadID())
 		}
 	}
-	if bp := trapthread.Breakpoint(); bp.Active {
-		return dbp.SwitchThread(trapthread.ThreadID())
+	if trapthread != nil {
+		if bp := trapthread.Breakpoint(); bp.Active {
+			return dbp.SwitchThread(trapthread.ThreadID())
+		}
 	}
 	for _, th := range threads {
 		if bp := th.Breakpoint(); bp.Active {
 			return dbp.SwitchThread(th.ThreadID())
 		}
 	}
-	return dbp.SwitchThread(trapthread.ThreadID())
+	if trapthread != nil {
+		return dbp.SwitchThread(trapthread.ThreadID())
+	}
+	if _, ok := dbp.FindThread(dbp.currentThread.ThreadID()); ok {
+		dbp.selectedGoroutine, _ = GetG(dbp.currentThread)
+		return nil
+	}
+	if len(threads) > 0 {
+		return dbp.SwitchThread(threads[0].ThreadID())
+	}
+	return nil
+}
+
+// pickCurrentTarget picks a new current target, with the following property:
+//
+//   - a target with an active stepping breakpoint
+//   - a target with StopReason == StopCallReturned
+//   - a target with an active breakpoint, prioritizing traptgt
+//   - traptgt
+func pickCurrentTarget(grp *TargetGroup, traptgt *Target) *Target {
```

Comment:
This feels like it should be a method instead of a function as it's use is to set `TargetGroup.Selected` and isn't really used outside of that, so it might as well set that directly and then callers can use `TargetGroup.Selected` directly after that. Or perhaps we also return the selected target for convenience but also set `TargetGroup.Selected` within the method.